src/Text.tsx

```diff
@@ -1,13 +1,53 @@
+import React from 'react'
 import styled from 'styled-components'
 import {COMMON, SystemCommonProps, SystemTypographyProps, TYPOGRAPHY} from './constants'
+import {ForwardRefComponent as PolymorphicForwardRefComponent} from './utils/polymorphic'
 import sx, {SxProp} from './sx'
-import {ComponentProps} from './utils/types'
 
-const Text = styled.span<SystemTypographyProps & SystemCommonProps & SxProp>`
+const restrictedAs = [
+  'span',
+  'p',
+  'b',
+  'i',
+  'em',
+  'small',
+  'strong',
+  'u',
+  'code',
+  'q',
+  's',
+  'label',
+  'ins',
+  'del',
+  'mark',
+  'kbd',
+  'abbr',
+] as const
+
+const StyledText = styled.span<SystemTypographyProps & SystemCommonProps & SxProp>`
   ${TYPOGRAPHY};
   ${COMMON};
   ${sx};
 `
 
-export type TextProps = ComponentProps<typeof Text>
+export type TextProps = {
+  as?: typeof restrictedAs[number]
+} & React.ComponentProps<'span' & typeof StyledText>
+
+const Text = React.forwardRef(({children, as = 'span', ...rest}, forwardedRef) => {
+  if (__DEV__ && !restrictedAs.includes(as)) {
+    console.warn(
```

Comment:
You can use an eslint-disable comment to hide this lint error: https://github.com/primer/react/blob/main/src/Link.tsx#L62